`CMakeSourceGenerate` step could intermittently fail while downloading the SDK source from a GitHub tag

Maye we want to have some retry logic here to be resilient to network related errors when we are downloading sources:
https://github.com/Azure/azure-sdk-for-cpp/blob/af6d5dcd076abe795285280da0dbfc3c535237b2/cmake-modules/FolderList.cmake#L79-L83

Here's the cmake command in question being used:
https://cmake.org/cmake/help/latest/command/file.html#transfer

We should also have some more verbose logging to help diagnose why such a command would fail, using `STATUS <variable>`. 

Intermittent error (retry fixed it):
https://dev.azure.com/azure-sdk/internal/_build/results?buildId=3322857&view=logs&j=0e8a4fef-2e3d-5eed-ec7c-86f1a97aad2d&t=c03eb77c-6421-5173-32b7-bd05776958ed&s=3fdee543-84bb-5239-0323-7c652c19379c

```text
2023-12-06T20:05:15.5215780Z ##[section]Starting: cmake generate for azure-identity
2023-12-06T20:05:15.5226460Z ==============================================================================
2023-12-06T20:05:15.5227080Z Task         : Command line
2023-12-06T20:05:15.5227440Z Description  : Run a command line script using Bash on Linux and macOS and cmd.exe on Windows
2023-12-06T20:05:15.5227820Z Version      : 2.231.0
2023-12-06T20:05:15.5228120Z Author       : Microsoft Corporation
2023-12-06T20:05:15.5228440Z Help         : https://docs.microsoft.com/azure/devops/pipelines/tasks/utility/command-line
2023-12-06T20:05:15.5228790Z ==============================================================================
2023-12-06T20:05:16.0842390Z Generating script.
2023-12-06T20:05:16.0965700Z Script contents:
2023-12-06T20:05:16.0984910Z OPENSSL_ROOT_DIR=/usr/local/opt/openssl@3 OPENSSL_INCLUDE_DIR=/usr/local/opt/openssl@3/include  cmake -DFETCH_SOURCE_DEPS=ON ..
2023-12-06T20:05:16.1006450Z ========================== Starting Command Output ===========================
2023-12-06T20:05:16.1801090Z [command]/bin/bash --noprofile --norc /Users/runner/work/_temp/0db1b6fb-60d8-411e-8ee7-c0c52eb9a84a.sh
2023-12-06T20:05:16.1904560Z Vcpkg integrate step.
2023-12-06T20:05:16.2010440Z CMAKE_TOOLCHAIN_FILE is not defined. Define it for the user.
2023-12-06T20:05:16.2114630Z AZURE_SDK_DISABLE_AUTO_VCPKG is not defined. Fetch a local copy of vcpkg.
2023-12-06T20:05:16.2230550Z Vcpkg commit string used: 43cf47eccfbe27006cf9534a5db809798f8c37fe
2023-12-06T20:05:51.6481270Z Vcpkg integrate step - DONE.
2023-12-06T20:05:51.6655050Z -- Bootstrapping vcpkg before install
2023-12-06T20:05:53.9526260Z -- Bootstrapping vcpkg before install - done
2023-12-06T20:05:53.9536930Z -- Running vcpkg install
2023-12-06T20:05:54.1836250Z Detecting compiler hash for triplet x64-osx...
2023-12-06T20:06:00.5264520Z The following packages will be built and installed:
2023-12-06T20:06:00.5277730Z     azure-core-cpp[core,curl,http]:x64-osx -> 1.10.3#2
2023-12-06T20:06:00.5380530Z   * curl[core,non-http,sectransp,ssl]:x64-osx -> 8.4.0
2023-12-06T20:06:00.5438410Z     openssl:x64-osx -> 3.1.4#1
2023-12-06T20:06:00.5541580Z   * vcpkg-cmake:x64-osx -> 2023-05-04
2023-12-06T20:06:00.5554290Z   * vcpkg-cmake-config:x64-osx -> 2022-02-06#1
2023-12-06T20:06:00.5656550Z   * vcpkg-cmake-get-vars:x64-osx -> 2023-03-02
2023-12-06T20:06:00.5759130Z   * zlib:x64-osx -> 1.3
2023-12-06T20:06:00.5861600Z Additional packages (*) will be modified to complete this operation.
2023-12-06T20:06:06.3630650Z Restored 7 package(s) from HTTP servers in 5.8 s. Use --debug to see more details.
2023-12-06T20:06:06.3632320Z Installing 1/7 vcpkg-cmake:x64-osx...
2023-12-06T20:06:06.3692390Z Elapsed time to handle vcpkg-cmake:x64-osx: 6.11 ms
2023-12-06T20:06:06.3696630Z Installing 2/7 zlib:x64-osx...
2023-12-06T20:06:06.3868420Z Elapsed time to handle zlib:x64-osx: 8.54 ms
2023-12-06T20:06:06.3870520Z Installing 3/7 vcpkg-cmake-config:x64-osx...
2023-12-06T20:06:06.3875830Z Elapsed time to handle vcpkg-cmake-config:x64-osx: 4.68 ms
2023-12-06T20:06:06.3877470Z Installing 4/7 curl:x64-osx...
2023-12-06T20:06:06.4070160Z Elapsed time to handle curl:x64-osx: 23.8 ms
2023-12-06T20:06:06.4098610Z Installing 5/7 vcpkg-cmake-get-vars:x64-osx...
2023-12-06T20:06:06.4116330Z Elapsed time to handle vcpkg-cmake-get-vars:x64-osx: 5.42 ms
2023-12-06T20:06:06.4118060Z Installing 6/7 openssl:x64-osx...
2023-12-06T20:06:06.4793080Z Elapsed time to handle openssl:x64-osx: 67.5 ms
2023-12-06T20:06:06.4796230Z Installing 7/7 azure-core-cpp:x64-osx...
2023-12-06T20:06:06.5141150Z Elapsed time to handle azure-core-cpp:x64-osx: 34.9 ms
2023-12-06T20:06:06.5142260Z Total install time: 151 ms
2023-12-06T20:06:06.5143950Z The package openssl is compatible with built-in CMake targets:
2023-12-06T20:06:06.5144410Z 
2023-12-06T20:06:06.5145010Z     find_package(OpenSSL REQUIRED)
2023-12-06T20:06:06.5150630Z     target_link_libraries(main PRIVATE OpenSSL::SSL OpenSSL::Crypto)
2023-12-06T20:06:06.5156710Z 
2023-12-06T20:06:06.5196850Z azure-core-cpp provides CMake targets:
2023-12-06T20:06:06.5198960Z 
2023-12-06T20:06:06.5200360Z   # this is heuristically generated, and may not be correct
2023-12-06T20:06:06.5201950Z   find_package(azure-core-cpp CONFIG REQUIRED)
2023-12-06T20:06:06.5203360Z   target_link_libraries(main PRIVATE Azure::azure-core)
2023-12-06T20:06:06.5204340Z 
2023-12-06T20:06:06.6128170Z -- Running vcpkg install - done
2023-12-06T20:06:08.1524410Z -- The CXX compiler identification is AppleClang 13.0.0.13000029
2023-12-06T20:06:08.2829290Z -- Detecting CXX compiler ABI info
2023-12-06T20:06:08.6578050Z -- Detecting CXX compiler ABI info - done
2023-12-06T20:06:08.6722090Z -- Check for working CXX compiler: /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/c++ - skipped
2023-12-06T20:06:08.6726670Z -- Detecting CXX compile features
2023-12-06T20:06:08.6738690Z -- Detecting CXX compile features - done
2023-12-06T20:06:08.6788680Z setting up compile options for IDENTITY
2023-12-06T20:06:08.6801620Z No transport adapter was selected, using libcurl as the default option for POSIX.
2023-12-06T20:06:08.6830670Z project found IDENTITY
2023-12-06T20:06:08.6832230Z FLAG VALUE : ON
2023-12-06T20:06:08.6833210Z Downloading version 1.2.0 of azure-core
2023-12-06T20:06:08.7230650Z tar: could not chdir to '/Users/runner/work/1/s/sdk/identity/azure-identity/build/downloads/azure-core/azure-sdk-for-cpp-azure-core_1.2.0/sdk/core'
2023-12-06T20:06:08.7233090Z 
2023-12-06T20:06:08.7234420Z add folder build/sdk/core
2023-12-06T20:06:08.7235850Z CMake Error at CMakeLists.txt:38 (add_subdirectory):
2023-12-06T20:06:08.7237130Z   The source directory
2023-12-06T20:06:08.7238030Z 
2023-12-06T20:06:08.7239440Z     /Users/runner/work/1/s/sdk/identity/azure-identity/build/sdk/core
2023-12-06T20:06:08.7240360Z 
2023-12-06T20:06:08.7241780Z   does not contain a CMakeLists.txt file.
2023-12-06T20:06:08.7242670Z 
2023-12-06T20:06:08.7243540Z 
2023-12-06T20:06:09.0730090Z -- Found OpenSSL: /Users/runner/work/1/s/sdk/identity/azure-identity/build/vcpkg_installed/x64-osx/debug/lib/libcrypto.a (found version "3.1.4")  
2023-12-06T20:06:09.0767400Z -- Retrieving version from /Users/runner/work/1/s/sdk/identity/azure-identity/src/private/package_version.hpp
2023-12-06T20:06:09.0784510Z -- VERSION_MAJOR 1
2023-12-06T20:06:09.0786120Z -- VERSION_MINOR 7
2023-12-06T20:06:09.0794700Z -- VERSION_PATCH 0
2023-12-06T20:06:09.0796250Z -- VERSION_PRERELEASE beta.1
2023-12-06T20:06:09.0797830Z -- AZ_LIBRARY_VERSION 
2023-12-06T20:06:09.0971570Z -- Configuring incomplete, errors occurred!
2023-12-06T20:06:09.1332400Z ##[error]Bash exited with code '1'.
2023-12-06T20:06:09.1533040Z ##[section]Finishing: cmake generate for azure-identity
```

cc @danieljurek, @gearama

--- a/cmake-modules/FolderList.cmake
+++ b/cmake-modules/FolderList.cmake
@@ -73,13 +73,13 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
     set(DEP_PREFIX azure-sdk-for-cpp)
 
     if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
+        SET(DOWNLOAD_MESSAGE "Downloading latest version of ${DEP_NAME}")
         #get the latest version from main
         SET(DOWNLOAD_LINK "http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip")
-        SET(DOWNLOAD_MESSAGE "Downloading latest version of ${DEP_NAME}")
     else()
+        SET(DOWNLOAD_MESSAGE "Downloading version ${DEP_VERSION} of ${DEP_NAME}")
         # get the zip
         SET(DOWNLOAD_LINK "https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE}")
-        SET(DOWNLOAD_MESSAGE "Downloading version ${DEP_VERSION} of ${DEP_NAME}")
     endif()
 
     foreach(RETRY_ATTEMPT RANGE 2)

--- a/cmake-modules/FolderList.cmake
+++ b/cmake-modules/FolderList.cmake
@@ -72,15 +72,34 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
     set(DOWNLOAD_FILE ${DEP_NAME}_${DEP_VERSION}.zip)
     set(DEP_PREFIX azure-sdk-for-cpp)
 
-    if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
-        message("Downloading latest version of ${DEP_NAME}")
-        #get the latest version from main
-        file(DOWNLOAD http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE})
-    else()
-        message("Downloading version ${DEP_VERSION} of ${DEP_NAME}")
-        # get the zip
-        file(DOWNLOAD https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE} ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE})
-    endif()
+    foreach(RETRY_ATTEMPT RANGE 2)
+        if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
+            message("Downloading latest version of ${DEP_NAME}")
+            #get the latest version from main
+            file(
+                DOWNLOAD http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip
+                ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE}
+                SHOW_PROGRESS
+                STATUS DOWNLOAD_STATUS
+            )
+        else()
+            message("Downloading version ${DEP_VERSION} of ${DEP_NAME}")
+            # get the zip
+            file(
+                DOWNLOAD https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE}
+                ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE}
+                SHOW_PROGRESS
+                STATUS DOWNLOAD_STATUS
+            )
+        endif()
+
+        list(GET DOWNLOAD_STATUS 0 STATUS_CODE)
+        if (${STATUS_CODE} EQUAL 0)
+            break()
+        else()
+            message("Download unsuccessful, retry attempt will be made.")
+        endif()
+    endforeach()
 
     #extract the zip
     file(ARCHIVE_EXTRACT INPUT ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE} DESTINATION ${DOWNLOAD_FOLDER}/${DEP_NAME})

--- a/cmake-modules/FolderList.cmake
+++ b/cmake-modules/FolderList.cmake
@@ -72,6 +72,16 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
     set(DOWNLOAD_FILE ${DEP_NAME}_${DEP_VERSION}.zip)
     set(DEP_PREFIX azure-sdk-for-cpp)
 
+    if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
+        #get the latest version from main
+        SET(DOWNLOAD_LINK "http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip")
+        SET(DOWNLOAD_MESSAGE "Downloading latest version of ${DEP_NAME}")
+    else()
+        # get the zip
+        SET(DOWNLOAD_LINK "https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE}")
+        SET(DOWNLOAD_MESSAGE "Downloading version ${DEP_VERSION} of ${DEP_NAME}")
+    endif()
+
     foreach(RETRY_ATTEMPT RANGE 2)
         math(EXPR RETRY_DELAY "10 * ${RETRY_ATTEMPT}" OUTPUT_FORMAT DECIMAL)
         if (RETRY_DELAY GREATER 0)
@@ -79,16 +89,7 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
             execute_process(COMMAND ${CMAKE_COMMAND} -E sleep ${RETRY_DELAY})
         endif()
 
-        if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
-            #get the latest version from main
-            SET(DOWNLOAD_LINK "http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip")
-            message("Downloading latest version of ${DEP_NAME}")
-        else()
-            # get the zip
-            SET(DOWNLOAD_LINK "https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE}")
-            message("Downloading version ${DEP_VERSION} of ${DEP_NAME}")
-        endif()
-    
+        message(${DOWNLOAD_MESSAGE})
         file(
             DOWNLOAD ${DOWNLOAD_LINK}
             ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE}

--- a/cmake-modules/FolderList.cmake
+++ b/cmake-modules/FolderList.cmake
@@ -73,31 +73,40 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
     set(DEP_PREFIX azure-sdk-for-cpp)
 
     foreach(RETRY_ATTEMPT RANGE 2)
-        if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
-            message("Downloading latest version of ${DEP_NAME}")
-            #get the latest version from main
-            file(
-                DOWNLOAD http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip
-                ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE}
-                SHOW_PROGRESS
-                STATUS DOWNLOAD_STATUS
-            )
-        else()
-            message("Downloading version ${DEP_VERSION} of ${DEP_NAME}")
-            # get the zip
-            file(
-                DOWNLOAD https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE}
-                ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE}
-                SHOW_PROGRESS
-                STATUS DOWNLOAD_STATUS
-            )
+        math(EXPR RETRY_DELAY "10 * ${RETRY_ATTEMPT}" OUTPUT_FORMAT DECIMAL)
+        if (RETRY_DELAY GREATER 0)
+            message("Waiting for ${RETRY_DELAY} seconds before retrying download.")
+            execute_process(COMMAND ${CMAKE_COMMAND} -E sleep ${RETRY_DELAY})
         endif()
+
+        if(FETCH_SOURCE_DEPS STREQUAL "LATEST")
+            #get the latest version from main
+            SET(DOWNLOAD_LINK "http://github.com/Azure/azure-sdk-for-cpp/archive/main.zip")
+            message("Downloading latest version of ${DEP_NAME}")
+        else()
+            # get the zip
+            SET(DOWNLOAD_LINK "https://github.com/Azure/azure-sdk-for-cpp/archive/refs/tags/${DOWNLOAD_FILE}")
+            message("Downloading version ${DEP_VERSION} of ${DEP_NAME}")
+        endif()
+    
+        file(
+            DOWNLOAD ${DOWNLOAD_LINK}
+            ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE}
+            SHOW_PROGRESS
+            STATUS DOWNLOAD_STATUS
+        )
 
         list(GET DOWNLOAD_STATUS 0 STATUS_CODE)
         if (${STATUS_CODE} EQUAL 0)
             break()
+        else()
+            list(GET DOWNLOAD_STATUS 1 ERROR_MESSAGE)
+            message("Download failed with status code ${STATUS_CODE}: ${ERROR_MESSAGE}.")
         endif()
     endforeach()
+    if (NOT ${STATUS_CODE} EQUAL 0)
+        message(FATAL_ERROR "Dependency download failed (Link: ${DOWNLOAD_LINK}).")
+    endif()
 
     #extract the zip
     file(ARCHIVE_EXTRACT INPUT ${DOWNLOAD_FOLDER}/${DOWNLOAD_FILE} DESTINATION ${DOWNLOAD_FOLDER}/${DEP_NAME})

--- a/cmake-modules/FolderList.cmake
+++ b/cmake-modules/FolderList.cmake
@@ -96,8 +96,6 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
         list(GET DOWNLOAD_STATUS 0 STATUS_CODE)
         if (${STATUS_CODE} EQUAL 0)
             break()
-        else()
-            message("Download unsuccessful, retry attempt will be made.")
         endif()
     endforeach()
 

--- a/cmake-modules/FolderList.cmake
+++ b/cmake-modules/FolderList.cmake
@@ -84,7 +84,7 @@ macro(DownloadDepVersion DEP_FOLDER DEP_NAME DEP_VERSION)
 
     foreach(RETRY_ATTEMPT RANGE 2)
         math(EXPR RETRY_DELAY "10 * ${RETRY_ATTEMPT}" OUTPUT_FORMAT DECIMAL)
-        if (RETRY_DELAY GREATER 0)
+        if (RETRY_ATTEMPT GREATER 0)
             message("Waiting for ${RETRY_DELAY} seconds before retrying download.")
             execute_process(COMMAND ${CMAKE_COMMAND} -E sleep ${RETRY_DELAY})
         endif()

